d2/src/main/java/com/linkedin/d2/discovery/stores/zk/ZooKeeperEphemeralStore.java

```diff
@@ -480,11 +490,33 @@ public void stopPublishing(String prop)
     }
   }
 
+  public String getConnectString() {
+    return _zkConn.getConnectString();
+  }
+
   public int getListenerCount()
   {
     return _useNewWatcher ? _ephemeralStoreWatchers.size() : _zkStoreWatcher.getWatchCount();
   }
 
+  public void setServiceDiscoveryEventHelper(D2ServiceDiscoveryEventHelper helper) {
+    _eventHelper = helper;
+  }
+
+  public void setZnodePathAndDataCallback(ZookeeperNodePathAndDataCallback callback) {
+    _znodePathAndDataCallback = callback;
+  }
+
+  public interface ZookeeperNodePathAndDataCallback {
+    void setPathAndDataForCluster(String cluster, String nodePath, String data);
```

Comment:
setPathAndDataForCluster : this does not seem to be used across this PR, could you document a bit about the purpose of this? My guess is that we want to this to get the ZK path of the announced server?

Reply:
It's used at line 555 in (store.setZnodePathAndDataCallback) in ZooKeeperAnnouncer.java

Reply:
Sounds good, these callbacks are crazy! (not your fault)